File: open_mahjong_server/server/gamestate/game_guobiao/wait_tips.py
"""Build MMCR-style winning-tile hints from the authoritative GB calculator."""

from collections import Counter
from typing import Dict, Iterable, List, Optional


FLOWERS = set(range(51, 59))


def _meld_tiles(target: str) -> List[int]:
    if not target or len(target) < 2:
        return []
    try:
        tile = int(target[1:])
    except (TypeError, ValueError):
        return []
    prefix = target[0]
    if prefix.lower() == "s":
        return [tile - 1, tile, tile + 1]
    if prefix == "k":
        return [tile] * 3
    if prefix in ("g", "G"):
        return [tile] * 4
    return []


def _table_counts(game_state) -> Counter:
    counts: Counter = Counter()
    for player in game_state.player_list:
        counts.update(int(tile) for tile in player.discard_tiles)
        for target in player.combination_tiles:
            counts.update(_meld_tiles(target))
    return counts


def _remaining_count(game_state, hand: Iterable[int], tile: int, pending_cut: Optional[int]) -> int:
    visible = _table_counts(game_state)
    visible.update(int(value) for value in hand)
    if pending_cut is not None:
        visible.update([pending_cut])
    return max(0, 4 - visible[tile])


def _base_way(game_state, player_index: int, wait_count: int) -> List[str]:
    player = game_state.player_list[player_index]
    way = ["花牌"] * len(player.huapai_list)
    round_index = int(getattr(game_state, "current_round", 1))
    round_winds = ("场风东", "场风南", "场风西", "场风北")
    way.append(round_winds[min(3, max(0, (round_index - 1) // 4))])
    seat_winds = ("自风东", "自风南", "自风西", "自风北")
    way.append(seat_winds[int(player.player_index) % 4])
    if wait_count == 1:
        way.append("和单张")
    return way


def _score(game_state, hand: List[int], combinations: List[str], way: List[str], tile: int) -> int:
    service = game_state.calculation_service
    try:
        sub_rule = getattr(game_state, "sub_rule", "guobiao/standard")
        if sub_rule == "guobiao/xiaolin":
            result = service.GB_xiaolin_hepai_check(list(hand), list(combinations), list(way), tile)
        elif sub_rule == "guobiao/kshen":
            result = service.GB_kshen_hepai_check(list(hand), list(combinations), list(way), tile)
        else:
            result = service.GB_hepai_check(list(hand), list(combinations), list(way), tile)
        return max(0, int(result[0]))
    except Exception:
        return 0


def _details_for_hand(
    game_state,
    player_index: int,
    hand: List[int],
    pending_cut: Optional[int] = None,
) -> List[Dict[str, object]]:
    player = game_state.player_list[player_index]
    combinations = list(player.combination_tiles)
    try:
        waiting = sorted(
            int(tile)
            for tile in game_state.calculation_service.GB_tingpai_check(list(hand), combinations)
        )
    except Exception:
        return []
    if not waiting:
        return []

    table = _table_counts(game_state)
    if pending_cut is not None:
        table.update([pending_cut])
    base_way = _base_way(game_state, player_index, len(waiting))
    flower_count = len(player.huapai_list)
    minimum = int(getattr(game_state, "hepai_limit", 8))
    details: List[Dict[str, object]] = []

    for tile in waiting:
        ron_way = list(base_way)
        if table[tile] == 4:
            ron_way.append("和绝张")
        ron_way.append("点和")

        tsumo_way = list(base_way)
        if table[tile] == 3:
            tsumo_way.append("和绝张")
        tsumo_way.append("自摸")

        complete_hand = list(hand) + [tile]
        ron_fan = _score(game_state, complete_hand, combinations, ron_way, tile)
        tsumo_fan = _score(game_state, complete_hand, combinations, tsumo_way, tile)
        ron_allowed = ron_fan - flower_count >= minimum
        tsumo_allowed = tsumo_fan - flower_count >= minimum
        details.append(
            {
                "tile": tile,
                "base_f": ron_fan if ron_allowed else 0,
                "selfdrawn_f": tsumo_fan if tsumo_allowed else 0,
                "remaining_count": _remaining_count(game_state, hand, tile, pending_cut),
            }
        )
    return details


def _build_wait_data(game_state, player_index: int, *, include_discards: bool) -> Optional[Dict[str, object]]:
    if not bool(getattr(game_state, "tips", False)):
        return None
    player = game_state.player_list[player_index]
    hand = list(player.hand_tiles)
    if any(tile in FLOWERS for tile in hand):
        return None

    if not include_discards:
        details = _details_for_hand(game_state, player_index, hand)
        return {"type": "waits", "details": details} if details else None

    by_discard: List[Dict[str, object]] = []
    seen = set()
    for index, discard in enumerate(hand):
        if discard in seen or discard in FLOWERS:
            continue
        seen.add(discard)
        remaining_hand = hand[:index] + hand[index + 1 :]
        adds = _details_for_hand(game_state, player_index, remaining_hand, pending_cut=discard)
        if adds:
            by_discard.append({"discard_tile": discard, "adds": adds})
    return {"type": "waits_all", "details": by_discard} if by_discard else None


def build_wait_data(game_state, player_index: int, *, include_discards: bool) -> Optional[Dict[str, object]]:
    """Return hints without ever allowing an optional UI calculation to stop play."""
    try:
        return _build_wait_data(game_state, player_index, include_discards=include_discards)
    except Exception:
        return None

File: open_mahjong_web/client/src/game2d/calc/guobiao/waitTips.ts
/**
 * Client-side guobiao wait tips, mirroring Unity TipsContainer + HeJuezhangTableCounter.
 * Output shape matches MMCR WaitDisplay (waits / waits_all).
 */

import {
  tingpaiCheck,
  hepaiCheck,
  hepaiCheckXiaolin,
  hepaiCheckKshen,
  type HepaiResult,
} from './index'

const FLOWERS = new Set([51, 52, 53, 54, 55, 56, 57, 58])

export interface WaitDetail {
  tile: number
  base_f: number
  selfdrawn_f: number
  remaining_count: number
}

export type WaitInfoData =
  | { type: 'waits'; details: WaitDetail[] }
  | { type: 'waits_all'; details: Array<{ discard_tile: number; adds: WaitDetail[] }> }
  | null

export interface WaitTipsContext {
  tips: boolean
  hand: number[]
  combinations: string[]
  flowerCount: number
  playerIndex: number
  currentRound: number
  hepaiLimit: number
  subRule: string
  /** Discard piles for seats 0-3 (Salasasa tile ids). */
  seatDiscards: number[][]
  /** Open meld keys for seats 0-3 (`s37`, `k39`, `g41`, …). */
  seatCombinations: string[][]
}

function scoreHepai(
  subRule: string,
  hand: number[],
  combinations: string[],
  way: string[],
  tile: number,
): number {
  let result: HepaiResult
  if (subRule === 'guobiao/xiaolin') {
    result = hepaiCheckXiaolin(hand, combinations, way, tile, false)
  } else if (subRule === 'guobiao/kshen') {
    result = hepaiCheckKshen(hand, combinations, way, tile, false)
  } else {
    result = hepaiCheck(hand, combinations, way, tile, false)
  }
  return Math.max(0, Number(result.fan) || 0)
}

function buildBaseWay(ctx: WaitTipsContext, waitCount: number): string[] {
  const way: string[] = []
  for (let i = 0; i < ctx.flowerCount; i += 1) way.push('花牌')
  const roundIndex = Math.max(1, Number(ctx.currentRound) || 1)
  if (roundIndex <= 4) way.push('场风东')
  else if (roundIndex <= 8) way.push('场风南')
  else if (roundIndex <= 12) way.push('场风西')
  else way.push('场风北')
  const seat = ((Number(ctx.playerIndex) % 4) + 4) % 4
  way.push(['自风东', '自风南', '自风西', '自风北'][seat])
  if (waitCount === 1) way.push('和单张')
  return way
}

/** C# HeJuezhangTableCounter: river + k/s melds only (no g/G), optional pending cut. */
function countShowTilesOnTable(
  hepaiTile: number,
  seatDiscards: number[][],
  seatCombinations: string[][],
  pendingCut: number | null,
): number {
  let count = 0
  for (const discards of seatDiscards) {
    for (const tile of discards) {
      if (tile === hepaiTile) count += 1
    }
  }
  for (const combinations of seatCombinations) {
    for (const combination of combinations) {
      if (!combination) continue
      if (combination.includes(`k${hepaiTile}`)) count += 3
      if (combination.includes(`s${hepaiTile - 1}`)) count += 1
      if (combination.includes(`s${hepaiTile}`)) count += 1
      if (combination.includes(`s${hepaiTile + 1}`)) count += 1
    }
  }
  if (pendingCut != null && pendingCut === hepaiTile) count += 1
  return count
}

function expandMeldTiles(target: string): number[] {
  if (!target || target.length < 2) return []
  const tile = Number(target.slice(1))
  if (!Number.isFinite(tile)) return []
  const prefix = target[0]
  if (prefix.toLowerCase() === 's') return [tile - 1, tile, tile + 1]
  if (prefix === 'k') return [tile, tile, tile]
  if (prefix === 'g' || prefix === 'G') return [tile, tile, tile, tile]
  if (prefix.toLowerCase() === 'q') return [tile, tile]
  return []
}

function remainingCount(
  hand: number[],
  seatDiscards: number[][],
  seatCombinations: string[][],
  tile: number,
  pendingCut: number | null,
): number {
  let used = 0
  for (const value of hand) {
    if (value === tile) used += 1
  }
  if (pendingCut != null && pendingCut === tile) used += 1
  for (const discards of seatDiscards) {
    for (const value of discards) {
      if (value === tile) used += 1
    }
  }
  for (const combinations of seatCombinations) {
    for (const combination of combinations) {
      for (const value of expandMeldTiles(combination)) {
        if (value === tile) used += 1
      }
    }
  }
  return Math.max(0, 4 - used)
}

function detailsForHand(
  ctx: WaitTipsContext,
  hand: number[],
  pendingCut: number | null = null,
): WaitDetail[] {
  let waiting: number[]
  try {
    waiting = tingpaiCheck(hand, ctx.combinations, false)
  } catch {
    return []
  }
  if (!waiting.length) return []

  const baseWay = buildBaseWay(ctx, waiting.length)
  const details: WaitDetail[] = []

  for (const tile of waiting) {
    const show = countShowTilesOnTable(tile, ctx.seatDiscards, ctx.seatCombinations, pendingCut)
    // Tips timing: and-tile not yet on table; show==3 ⇒ 和绝张 (C# TipsContainer).
    const hejuezhang = show === 3 ? ['和绝张'] : []
    const completeHand = [...hand, tile]

    const ronWay = [...baseWay, ...hejuezhang, '点和']
    const tsumoWay = [...baseWay, ...hejuezhang, '自摸']

    let ronFan = 0
    let tsumoFan = 0
    try {
      ronFan = scoreHepai(ctx.subRule, completeHand, ctx.combinations, ronWay, tile)
      tsumoFan = scoreHepai(ctx.subRule, completeHand, ctx.combinations, tsumoWay, tile)
    } catch {
      ronFan = 0
      tsumoFan = 0
    }

    const ronAllowed = ronFan - ctx.flowerCount >= ctx.hepaiLimit
    const tsumoAllowed = tsumoFan - ctx.flowerCount >= ctx.hepaiLimit
    details.push({
      tile,
      base_f: ronAllowed ? ronFan : 0,
      selfdrawn_f: tsumoAllowed ? tsumoFan : 0,
      remaining_count: remainingCount(hand, ctx.seatDiscards, ctx.seatCombinations, tile, pendingCut),
    })
  }
  return details
}

/**
 * Build wait tip payload for WaitDisplay. Never throws.
 * When includeDiscards is true (can cut), returns waits_all keyed by discard tile.
 */
export function buildLocalWaitData(
  ctx: WaitTipsContext,
  options: { includeDiscards: boolean },
): WaitInfoData {
  try {
    if (!ctx.tips) return null
    if (ctx.hand.some((tile) => FLOWERS.has(tile))) return null

    if (!options.includeDiscards) {
      const details = detailsForHand(ctx, ctx.hand)
      return details.length ? { type: 'waits', details } : null
    }

    const byDiscard: Array<{ discard_tile: number; adds: WaitDetail[] }> = []
    const seen = new Set<number>()
    for (let index = 0; index < ctx.hand.length; index += 1) {
      const discard = ctx.hand[index]
      if (seen.has(discard) || FLOWERS.has(discard)) continue
      seen.add(discard)
      const remainingHand = [...ctx.hand.slice(0, index), ...ctx.hand.slice(index + 1)]
      const adds = detailsForHand(ctx, remainingHand, discard)
      if (adds.length) byDiscard.push({ discard_tile: discard, adds })
    }
    return byDiscard.length ? { type: 'waits_all', details: byDiscard } : null
  } catch {
    return null
  }
}
